### DISCUSSION TITLE
Request: FontAwesome 5 & FontAwesome Pro

### DISCUSSION THREAD

**Discussion by ShakataGaNai** [ryanoasis/nerd-fonts#727] [Ideas]
#### 🎯 Subject of the issue

I'd like to put in a request for support of FontAwesome 5 and FontAwesome Pro. The later requires users licenses, so perhaps just support in the patcher.

**Comment by ryanoasis** [comment]
@ShakataGaNai Yes this is a very big idea! :+1:  I am aware of them but haven't done any movement on that front. :sweat_smile:

**Comment by ryanoasis** [comment]
I think we can probably rule out FA Pro (I don't feel confident exploring that route -- unless someone can convince me otherwise).

As for FA 5 perhaps we can give an option with the patcher, something like either FA 5 or the Fork Awesome (this was brought up in #241). FA 4.7 fork that is arguably more open.

I think we need to discuss pros/cons of each. I, myself, would need to read up on the direction of each project. Perhaps we need to choose one over the other, or at least one as the default. 

Usually we try not to care about a few duplicates between sets here and there.. however here I think it would be overkill to support both out of the box (in the released patched fonts).

**Comment by max-baz** [comment]
> Usually we try not to care about a few duplicates between sets here and there.. however here I think it would be overkill to support both out of the box (in the released patched fonts).

Given that FA 5 has new design for many icons, maybe it's not that bad idea to have both FA5 and ForkAwesome (based on FA 4) in the released fonts — there will be two icons for everything, but they will have different design after all. I've heard people who like the old design more.

Still, I'd like to see FA 5 be the "default" font, and let its icons stay in the codepoints currently allocated for FA. ForkAwesome should be put in a new location within Nerd Fonts.

**Comment by bvgastel** [comment]
Any news about this? I would really like the new icons in v5.

**Comment by aiguofer** [comment]
I'd love to use https://fontawesome.com/icons/memory?style=solid; after a while of digging around I found the reason that symbol is not available is because it was released in FA 5. Would be really interested in this!

**Comment by brennanfee** [comment]
Given the open-source focus of Nerd Fonts (especially from a redistribution perspective), my vote would be a switch away from FontAwesome and toward ForkAwesome.  If for no other reason than at least we would be on slid ground from a licensing and future access level.  Who knows if\when FontAwesome will make a\another profit driven decision to limit further the images they provide or the manner in which they can be used (non-commercial only is a common tactic).

Given that ForkAwesome was created expressly to be free\open and intends to stay that way it would make more sense to shift over to using that instead.

**Comment by kowalski7cc** [comment]
I think FontAwesome has way more icons (event in the free version) than ForkAwesome (1500+ vs 700+).
And I don't find FontAwesome license so restrictive (CC BY 4.0)

**Comment by TechieAndroid** [comment]
I've tried asking on here and haven't gotten any answer yet but I was pointed here to ask.

Is it because font awesome hasn't been updated in nerd fonts that none of the glyphs work?
https://fontawesome.com/cheatsheet

**Comment by Soooda** [comment]
Just spent the whole night and found out FA 5 has not been included yet xD
I would definitely vote for updating to FA5. PLEASE!

**Comment by Erothas** [comment]
After two days of trying to figure out why my status bar icons break whenever I install nerd fonts, I finally figured out that it's because the damn awesome icons I use are not updated in nerd fonts. It has been over THREE YEARS since this issue was started. At this point, the owner will just not do it for whatever reason that I could not care less about. And I would honestly be more angry if I give up now after wasting two whole days for no reason.

So, Is there any way to do this manually? If anyone could please direct me towards the right way so we can do it ourselves. That would be great.

**Comment by paulhersch** [comment]
Same as Erothas, commenting for notification

**Comment by ShakataGaNai** [comment]
FYI - On the right hand side of this page (and any issue on GitHub) there is a section labeled "Notification" with a big green "Subscribe" button. That way you can tune into an issue without spamming.

**Comment by ryanoasis** [comment]
converting this from an issue to a discussion

**Comment by earboxer** [comment]
I would certainly prefer Fork-Awesome as it seems more "community"-minded (accepting external contributions, icons for popular open source projects). I think Font Awesome might be "changing the open-source recipe" in their upcoming version 6.

Either way we go, we'd like to have codepoint compatibility with Font Awesome 4 so it's not *too* big of a breaking change. (Once we choose a path, probably won't be possible to backtrack).

* Font Awesome v4 is F000-F2E0 (with some gaps) 
* **Fork** Awesome is 811 icons, F000-F372 (with some gaps)  (92% density)
   * We would only need to move Font Logos (Currently at F300-F32D) to keep the codepoints
* Font Awesome v5 Free is 1458 icons, E005, E007, E113, E01A, E01E, E041 - E088, F000 - F8FF. (61% density for the two regions)
   * We would need to move pretty much everything if we wanted to keep the codepoints.

Because of the sparseness of Font Awesome v5 Free, we probably wouldn't want to use its assigned codepoints for the non-v4 regions. If we're not using their new codepoints (arguably their largest asset, being an official successor to Font Awesome v4), then we would only choose based on the quantity or the quality of the icons.

If we were to go with Fork Awesome, and keep their codepoints then what to do with Font Logos? (unfortunately, the codepoints conflict with ours: a bunch of logos would get mixed up if we made this switch)...

(Personally, I'm not too concerned about updating, since material icons covers most of my needs).

**Reply by TechieAndroid** [reply]
I would be one to actually try to merge it in if I knew how to manage the codepoints of fonts.

Know of any good resources to learn how to start doing that?

Even if it was just a local repo I forked and kept updated. Or in the case of FA5 being copyright I could just create a walkthrough for others that want the bulk of the glyphs.

**Reply by Finii** [reply]
@TechieAndroid 
Patch in, with `font-patcher`?

Open the script in an editor and look for the `patch_set` definition:
![image](https://user-images.githubusercontent.com/16012374/151427914-34b0a483-a9aa-4d41-b5db-4023e7dad999.png)

The lines look like this:
```python
             {'Enabled': self.args.fontawesome,       
              'Name': "Font Awesome",          
              'Filename': "font-awesome/FontAwesome.otf",                 
              'Exact': True,                              
              'SymStart': 0xF000,
              'SymEnd': 0xF2E0,
              'SrcStart': None, 
              'SrcEnd': None, 
              'ScaleGlyph': FONTA_SCALE_LIST,
              'Attributes': SYM_ATTR_FONTA  
             }, 
```

You can 'grab' the existing Font Awesome line, or add another new one. A new one, is turned 'on' with `'Enabled': True,`.
Adjust the filename (it is relative to `src/glyphs`)
If `Exact` is `False` gaps in the symbol font are skipped (vanish in the patched font).
`SymStart` and `SymEnd` are the codepoints in the symbol fonts you want to patch in.
`SrcStart` and possibly `SrcEnd` are the destination codepoints in the patched font. If start is `None` it is the same as `SymStart`.
`ScaleGlyph` I would set to `None`
`Attributes` you probably want to set to `SYM_ATTR_DEFAULT`. This allows for special scaling rules for individual glyphs.


I assume you already did some self-patching? If not
* Download/extract the [font-patcher zip](https://github.com/ryanoasis/nerd-fonts/releases/download/2.2.0-RC/FontPatcher.zip).
* Put the new symbol font in the `src/glyphs` subdir.
* Put the to-be-patched font in the dir where `font-patcher` is.
* Run `fontforge --script font-patcher --complete name_of_to-be-patched_font.ttf`

**Reply by phdpe** [reply]
Finii,

I am trying to manually patch a font with Font Awesome 6 Free file. 

Here is what I have: 

* FontPatcher 2.2.0 RC downloaded from NerdFont repo
* `SF-Mono-Regular.otf` from Apple
* `Font Awesome 6 Free.otf` from Font Awesome (rename to `fa6.otf` and put to `src/glyphs` folder)

I first edited the `font-patcher` file, and changed the `custom` section to the following:

```
             {'Enabled': self.args.custom,       
              'Name': "Custom",          
              'Filename': "fa6.otf",                 
              'Exact': True,                              
              'SymStart': 0xF000,
              'SymEnd': 0xF5C8,
              'SrcStart': None, 
              'SrcEnd': None, 
              'ScaleGlyph': None,
              'Attributes': SYM_ATTR_DEFAULT  
             }, 
```

The reason I used `0xF5C8` as end codepoint is that i tested that was the last codepoint fa6 was able to show in my system.  

I then ran 

`./font-patcher SF-Mono-Regular.otf --custom fa6.otf`

The output, however, suggested that only 212 glyphs were patched, which cant be true because fa6 should contain about 1700 free icons.  

I also tried to change `SrcStart` and `SrcEnd` to match the ones in `SymStart` and `SymEnd`, and also change `Exact` to false, but the output was the same.

Can you help me figure out what I did wrong? 

Thank you!

**Reply by Finii** [reply]
Can reproduce. Examining.

**Reply by Finii** [reply]
The file you mentioned is not in the release package, as far as I can see:

![image](https://user-images.githubusercontent.com/16012374/155083110-a3673851-8915-4ed7-a28a-7b720bd677c6.png)

So I assumed you meant `Regular`, which ends up with mere 212 glyphs.
But then, on opening the fonts, the reason is clear... `Regular` has a vast 'underfill':

![image](https://user-images.githubusercontent.com/16012374/155083323-4ecfd12c-9e5b-4c73-a683-90bd0f5e4a66.png)

When using `Solid` there seem to be 824 glyphs in that range (`0xF5C8 - 0xF000` = 1480), which is still only 'half full', but it indeed is:
![image](https://user-images.githubusercontent.com/16012374/155083867-08e9cfe3-66e1-4744-9cd3-1694d303edaa.png)

So this kind of works. Still I see a problem when `'Exact': True`, which is probably a bug I introduced :unamused: Investigating.

**Reply by Finii** [reply]
> Still I see a problem when 'Exact': True, which is probably a bug I introduced

Ah, :sweat_smile: the odd behavior has always been there; it is just that no other source font (yet) has been designed like FA6, so that it was not detectable.

FA6 uses alternative codepoints a lot, and `font-patcher` does not handle that very gracefully.

So maybe (for experiments at least) you should use `'Exact': False`, because else the glyphs are added in their _primary_ location, which is not in the `0xF000` region at all. We put them in the codepoint that they themself declare to be; but if they are present multiple times in the symbol font that is not always the expected codepoint.

```
Updating glyph: <Glyph martini-glass-empty in font FontAwesome6Free-Solid> martini-glass-empty putting at: F000
ALTUNI for F000 (calls itself F000) : None

Updating glyph: <Glyph music in font FontAwesome6Free-Solid> music putting at: 1F3B5
ALTUNI for F001 (calls itself 1F3B5) : [['F001', -1, 0]]

Updating glyph: <Glyph magnifying-glass in font FontAwesome6Free-Solid> magnifying-glass putting at: 1F50D
ALTUNI for F002 (calls itself 1F50D) : [['F002', -1, 0]]

Updating glyph: <Glyph envelope in font FontAwesome6Free-Solid> envelope putting at: 1F582
ALTUNI for F003 (calls itself 1F582) : [['F0E0', -1, 0], ['F003', -1, 0], ['2709', -1, 0]]

Updating glyph: <Glyph heart in font FontAwesome6Free-Solid> heart putting at: 1F9E1
ALTUNI for F004 (calls itself 1F9E1) : [['1F90E', -1, 0], ['1F90D', -1, 0], ['1F5A4', -1, 0], ['1F49C', -1, 0], ['1F49B', -1, 0], ['1F49A', -1, 0], ['1F499', -1, 0], ['F08A', -1, 0], ['F004', -1, 0], ['2764', -1, 0], ['2665', -1, 0]]

Updating glyph: <Glyph star in font FontAwesome6Free-Solid> star putting at: F006
ALTUNI for F005 (calls itself F006) : [['F005', -1, 0], ['2B50', -1, 0]]

Updating glyph: <Glyph star in font FontAwesome6Free-Solid> star putting at: F006
ALTUNI for F006 (calls itself F006) : [['F005', -1, 0], ['2B50', -1, 0]]

Updating glyph: <Glyph user in font FontAwesome6Free-Solid> user putting at: 1F464
ALTUNI for F007 (calls itself 1F464) : [['F2C0', -1, 0], ['F007', -1, 0]]

Updating glyph: <Glyph film in font FontAwesome6Free-Solid> film putting at: 1F39E
ALTUNI for F008 (calls itself 1F39E) : [['F008', -1, 0]]

Updating glyph: <Glyph table-cells-large in font FontAwesome6Free-Solid> table-cells-large putting at: F009
ALTUNI for F009 (calls itself F009) : None
```
Fixed with #793

**Comment by Finii** [comment]
_Hmm, mentions seem to be not shown in discussions?_

This Discussion mentioned in #789 (Awesome v6) ...

**Reply by Finii** [reply]
See
* #1550
